src/osp/OneStepProverHostIo.sol

```diff
@@ -158,6 +177,47 @@ contract OneStepProverHostIo is IOneStepProver {
                 preimageEnd = preimage.length;
             }
             extracted = preimage[preimageOffset:preimageEnd];
+        } else if (inst.argumentData == 2) {
+            // The machine is asking for an Ethereum versioned hash preimage
+
+            require(proofType == 0, "UNKNOWN_PREIMAGE_PROOF");
+
+            bytes calldata kzgProof = proof[proofOffset:];
+
+            require(bytes32(kzgProof[:32]) == leafContents, "KZG_PROOF_WRONG_HASH");
+
+            (bool success, bytes memory kzgParams) = address(0x0A).staticcall(kzgProof);
+            require(success, "INVALID_KZG_PROOF");
+            require(kzgParams.length > 0, "KZG_PRECOMPILE_MISSING");
+            (uint256 fieldElementsPerBlob, ) = abi.decode(kzgParams, (uint256, uint256));
+
+            // If preimageOffset is greater than the blob size, leave extracted empty and call it here.
+            if (preimageOffset < fieldElementsPerBlob * 32) {
+                // We need to compute what point the polynomial should be evaluated at to get the right part of the preimage.
+                // KZG commitments use a bit reversal permutation to order the roots of unity.
+                // To account for that, we reverse the bit order of the index.
+                uint256 bitReversedIndex = 0;
+                // preimageOffset was required to be 32 byte aligned above
+                uint256 tmp = preimageOffset / 32;
+                for (uint256 i = 1; i < fieldElementsPerBlob; i <<= 1) {
+                    bitReversedIndex <<= 1;
+                    if (tmp & 1 == 1) {
+                        bitReversedIndex |= 1;
+                    }
+                    tmp >>= 1;
+                }
+
+                uint256 rootOfUnityPower = (1 << 32) / fieldElementsPerBlob;
+                rootOfUnityPower *= bitReversedIndex;
+                uint256 rootOfUnity = modExp256(
+                    PRIMITIVE_ROOT_OF_UNITY,
+                    rootOfUnityPower,
+                    BLS_MODULUS
```

Comment:
Why do we use a hardcoded BLS_MODULUS instead of the one returned from the precompile?

Reply:
Fixed, I think this code was originally before the precompile call which is why it was hardcoded